google-http-client-jackson2/src/test/java/com/google/api/client/json/jackson2/BatchTestWithFakeServer.java

```diff
@@ -0,0 +1,183 @@
+import com.google.api.services.storage.Storage;
+import com.google.api.services.storage.model.StorageObject;
+import com.sun.net.httpserver.Headers;
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.junit.Test;
+
+/**
+ * Test to verify https://github.com/apache/beam/pull/14527#discussion_r613980011.
+ *
+ * <p>I wanted to put this in google-http-client module, but google-http-client-json dependency
+ * would create a dependency cycle. Therefore I place this in this class.
+ */
+public class BatchTestWithFakeServer {
+
+  @Test
+  public void testErrorContentReadRetry_withFakeServer() throws IOException {
+    final int statusCode429_TooManyRequest = 429;
+
+    final HttpHandler handler =
+        new HttpHandler() {
+          int count = 0;
+
+          @Override
+          public void handle(HttpExchange httpExchange) throws IOException {
+            Headers responseHeaders = httpExchange.getResponseHeaders();
+            if (count == 0) {
+              // 1st request
+              byte[] response = "rateLimitExceeded".getBytes(StandardCharsets.UTF_8);
+              responseHeaders.set("Content-Type", "text/plain;  charset=UTF-8");
+              httpExchange.sendResponseHeaders(statusCode429_TooManyRequest, response.length);
+              try (OutputStream out = httpExchange.getResponseBody()) {
+                out.write(response);
+              }
+              count++;
+            } else {
+              // 2nd request
+              String contentBoundary = "batch_foobarbaz";
+              String contentBoundaryLine = "--" + contentBoundary;
+              String endOfContentBoundaryLine = "--" + contentBoundary + "--";
+              String content =
+                  contentBoundaryLine
+                      + "\n"
+                      + "Content-Type: application/http\n"
+                      + "\n"
+                      + "HTTP/1.1 404 Not Found\n"
+                      + "Content-Length: -1\n"
+                      + "\n"
+                      + "{\"error\":{\"code\":404}}"
+                      + "\n"
+                      + "\n"
+                      + endOfContentBoundaryLine
+                      + "\n";
+              byte[] response = content.getBytes(StandardCharsets.UTF_8);
+              responseHeaders.set("Content-Type", "multipart/mixed; boundary=" + contentBoundary);
+              httpExchange.sendResponseHeaders(200, response.length);
+              try (OutputStream out = httpExchange.getResponseBody()) {
+                out.write(response);
+              }
+            }
+          }
+        };
+    try (FakeServer server = new FakeServer(handler)) {
```

Comment:
With the fake server, this test succeeds. This implies that the mock setup in Beam is not enough (at least for this test case).